### PR TITLE
refactor(db): WIP new database API

### PR DESCRIPTION
**Pull request by davideast**
See #1158 for details. If you have comments about the API please direct your feedback to that issue to keep the discussion in one place. However, if you have questions or improvements for the code please place them here.

### CODE REVIEW

File: src/database/list/loaded.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { DatabaseQuery, ChildEvent, DatabaseSnapshot } from '../interfaces';
+import { fromRef } from '../observable/fromRef';
+import { createListSnapshotChanges } from './snapshot-changes';
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/operator/skipUntil';
+import 'rxjs/add/operator/skipWhile';
+import 'rxjs/add/operator/withLatestFrom';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/filter';
```
Comment (Toxicable) [inline src/database/list/loaded.ts]: Refactor these to `call` style?
such as mentioned here https://github.com/angular/material2/blob/c564df61f48ff9c40eea237c394a15adac19cfdb/CODING_STANDARDS.md#rxjs

Comment (davideast) [reply src/database/list/loaded.ts]: Yep! Totally on the `//TODO:  FIX THIS` list. Wanted to get the POC up and running first.

Comment (Toxicable) [reply src/database/list/loaded.ts]: Yeah no probs, just being nit picky

Comment (davideast) [reply src/database/list/loaded.ts]: Comments are always welcome :)

File: src/database/observable/fromRef.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  return new Observable<SnapshotPrevKey | null | undefined>(subscriber => {
+    const fn = ref[listenType](event, (snapshot, prevKey) => {
+      subscriber.next({ snapshot, prevKey })
+    }, subscriber.error.bind(subscriber));
+    return { unsubscribe() { ref.off(event, fn)} }
+  })
+  .map((combined: SnapshotPrevKey) =>  { 
+    const {snapshot, prevKey} = combined; 
+    return { event, snapshot, prevKey } 
+  }).delay(0);
```
[the remark this answers is not in the document]
Comment (davideast) [reply src/database/observable/fromRef.ts]: Yep! This is really for operations that happen with disconnected connections. They can sometimes happen synchronously, which is an unfortunate quirk of the SDK.

[the remark this answers is not in the document]
Comment (Toxicable) [reply src/database/observable/fromRef.ts]: Gotcha, sounds legit. Maybe a small comment noting that?

**Pull request merged**